### PR TITLE
crash fix and opt warning fix(make opt apply)

### PR DESCRIPTION
**Pull request by nouiz**
fix gh-5907

### CODE REVIEW

**Comment by aseyboldt** [comment]
Thanks for the fix. I can confirm that it works.

File: theano/tensor/opt.py [hunk 1]
```diff
@@ -5696,11 +5696,24 @@ def local_opt_alloc(node):
                                                     only_process_constants=True)
                     assert val.size == 1
                     # check which type of op
-                    casted = T.mul(*shapes).astype(str(input.dtype))
+                    size = T.mul(*shapes)
+                    if input.dtype == "float32":
+                        # shapes are ints and normally int64.
+                        # We don't want to have a float64 upcast here
+                        # if input is a float32.
+                        size = size.astype(input.dtype)
```
Comment (lamblin) [inline theano/tensor/opt.py]: Why to this cast here when we reduced on all axes, but not below when we reduce on only some of them?

Comment (nouiz) [inline theano/tensor/opt.py]: I don't understand your comment.

Comment (lamblin) [inline theano/tensor/opt.py]: The "if" on line 5692 means we are in a case where the reduction was on all axes. The "else" block after line 5721 is the case where the reduction is only done on some axes.
The code in both blocks is similar, but the the cast `size = size.astype(input.dtype)` only happens in the first block, here, and not in the other block (for instance, just after line 5730).
Why that difference?

Comment (nouiz) [inline theano/tensor/opt.py]: Good catch. I refactored the code to have that part comment to both branch.

**Pull request merged**